src/dragonball/src/vm/mod.rs

```diff
@@ -89,8 +144,659 @@ impl Default for VmConfigInfo {
+    pub fn resume_all_vcpus_with_downtime(&mut self) -> std::result::Result<(), VcpuManagerError> {
+        self.vcpu_manager()?.resume_all_vcpus()?;
+
+        Ok(())
+    }
+
+    pub(crate) fn init_devices(
+        &mut self,
+        epoll_manager: EpollManager,
+    ) -> std::result::Result<(), StartMicroVmError> {
+        info!(self.logger, "VM: initializing devices ...");
+
+        let com1_sock_path = self.vm_config.serial_path.clone();
+        let kernel_config = self
+            .kernel_config
+            .as_mut()
+            .ok_or(StartMicroVmError::MissingKernelConfig)?;
+
+        info!(self.logger, "VM: create interrupt manager");
+        self.device_manager
+            .create_interrupt_manager()
+            .map_err(StartMicroVmError::DeviceManager)?;
+
+        info!(self.logger, "VM: create devices");
+        let vm_as =
+            self.address_space
+                .get_vm_as()
+                .ok_or(StartMicroVmError::AddressManagerError(
+                    AddressManagerError::GuestMemoryNotInitialized,
+                ))?;
+        self.device_manager.create_devices(
+            vm_as.clone(),
+            epoll_manager,
+            kernel_config,
+            com1_sock_path,
+            self.dmesg_fifo.take(),
+            self.address_space.address_space(),
+        )?;
+
+        info!(self.logger, "VM: start devices");
+        self.device_manager.start_devices()?;
+
+        info!(self.logger, "VM: initializing devices done");
+        Ok(())
+    }
+
+    /// Remove devices when shutdown vm
+    pub fn remove_devices(&mut self) -> std::result::Result<(), StopMicrovmError> {
+        info!(self.logger, "VM: remove devices");
+        let vm_as = self
+            .address_space
+            .get_vm_as()
+            .ok_or(StopMicrovmError::GuestMemoryNotInitialized)?;
+
+        self.device_manager
+            .remove_devices(
+                vm_as.clone(),
+                self.epoll_manager.clone(),
+                self.address_space.address_space(),
+            )
+            .map_err(StopMicrovmError::DeviceManager)
+    }
+
+    /// Reset the console into canonical mode.
+    pub fn reset_console(&self) -> std::result::Result<(), DeviceMgrError> {
+        self.device_manager.reset_console()
+    }
+
+    pub(crate) fn init_dmesg_logger(&mut self) {
+        let writer = self.dmesg_logger();
+        self.dmesg_fifo = Some(writer);
+    }
+
+    /// dmesg write to logger
+    fn dmesg_logger(&self) -> Box<dyn io::Write + Send> {
+        Box::new(DmesgWriter::new(&self.logger))
+    }
+
+    pub(crate) fn init_guest_memory(&mut self) -> std::result::Result<(), StartMicroVmError> {
+        info!(self.logger, "VM: initializing guest memory...");
+        // We are not allowing reinitialization of vm guest memory.
+        if self.address_space.is_initialized() {
+            return Ok(());
+        }
+
+        // vcpu boot up require local memory. reserve 100 MiB memory
```

Comment:
ooi, is this value generic or does it depend on architecture?

Reply:
We found that trait in both x86_64 and aarch64 .